File: 213.house_robber_ii/rob2.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {
    var len = nums.length;
    if (len === 0) return 0;
    if (len === 1) return nums[0];
    var maxValue = [];
    //do not use the first element
    maxValue[0] = 0;
    maxValue[1] = nums[1];
    for(var i = 2; i < len; i++){
        maxValue[i] = Math.max(maxValue[i-1], maxValue[i-2] + nums[i]);
    }

    var answer_without_first_elment = maxValue[len-1];
    //include the first element
    maxValue[0] = nums[0];
    maxValue[1] = Math.max(nums[0], nums[1]);
    for(var j = 2; j < len; j++){
        maxValue[j] = Math.max(maxValue[j-1], maxValue[j-2] + nums[j]);
    }
    return Math.max(maxValue[len-2],answer_without_first_elment);//use len-2 to eleminate the last one
};

console.log(rob([1,2,3,4,5,6]));

File: 213.house_robber_ii/rob2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums:
            return 0
        if len(nums) <= 2:
            return max(nums)
        # rob the first one
        dp = [[0, 0] for i in range(len(nums))]
        for i in range(2, len(nums)-1):
            # skip this house
            dp[i][0] = max(dp[i-1][0], dp[i-1][1])
            # rob this house
            dp[i][1] = dp[i-1][0] + nums[i]

        rob_first = nums[0] + max(dp[-2][0], dp[-2][1])
        # don't rob the first
        dp = [[0, 0] for i in range(len(nums))]
        for i in range(1, len(nums)):
            # skip this house
            dp[i][0] = max(dp[i-1][0], dp[i-1][1])
            # rob this house
            dp[i][1] = dp[i-1][0] + nums[i]
        not_rob_first = max(dp[-1][0], dp[-1][1])
        return max(rob_first, not_rob_first)


'''
use dp to record 2 cases:
    - rob the first house
          - skip the last
    - don't rob the first 
          - skip the first
'''
